e2e/rbd.go

```diff
@@ -55,104 +55,103 @@ var (
+	rbdDirPath + rbdProvisionerRBAC,
+	rbdDirPath + rbdProvisionerPSP,
+	rbdDirPath + rbdNodePlugin,
+	rbdDirPath + rbdNodePluginRBAC,
+	rbdDirPath + rbdNodePluginPSP,
 }
 
-func createORDeleteRbdResources(action string) {
-	csiDriver, err := ioutil.ReadFile(rbdDirPath + csiDriverObject)
-	if err != nil {
-		// createORDeleteRbdResources is used for upgrade testing as csidriverObject is
-		// newly added, discarding file not found error.
-		if !os.IsNotExist(err) {
-			e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+csiDriverObject, err)
-		}
-	} else {
-		_, err = framework.RunKubectlInput(cephCSINamespace, string(csiDriver), action, "-f", "-")
-		if err != nil {
-			e2elog.Failf("failed to %s CSIDriver object with error %v", action, err)
-		}
-	}
-	data, err := replaceNamespaceInTemplate(rbdDirPath + rbdProvisioner)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisioner, err)
-	}
-	data = oneReplicaDeployYaml(data)
-	data = enableTopologyInTemplate(data)
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
-	if err != nil {
-		e2elog.Failf("failed to %s rbd provisioner with error %v", action, err)
-	}
+type rbdDeploy struct{}
 
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdProvisionerRBAC)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisionerRBAC, err)
-	}
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, ns, "-f", "-")
+func (rd *rbdDeploy) replaceNamespaceInTemplate(filePath string)(string, error) {
+	read, err := ioutil.ReadFile(filePath)
 	if err != nil {
-		e2elog.Failf("failed to %s provisioner rbac with error %v", action, err)
+		return "", err
 	}
+	return strings.ReplaceAll(string(read), "namespace: default", fmt.Sprintf("namespace: %s", cephCSINamespace)), nil
+}
 
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdProvisionerPSP)
+func (rd *rbdDeploy) runAction(Template string, action string) {
+	data, err := rd.replaceNamespaceInTemplate(Template)
 	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdProvisionerPSP, err)
+		e2elog.Failf("failed to read content from %s with error %v", Template, err)
 	}
-	_, err = framework.RunKubectlInput(cephCSINamespace, data, action, "-f", "-")
+	_, err = framework.RunKubectlInput(cephCSINamespace, data, "--ignore-not-found=true", ns, action, "-f", "-")
 	if err != nil {
-		e2elog.Failf("failed to %s provisioner psp with error %v", action, err)
+		e2elog.Failf("failed to %s '%s' with error %v", action, Template, err)
 	}
+}
 
-	data, err = replaceNamespaceInTemplate(rbdDirPath + rbdNodePlugin)
-	if err != nil {
-		e2elog.Failf("failed to read content from %s with error %v", rbdDirPath+rbdNodePlugin, err)
+func (rd *rbdDeploy) changeTemplateAndRunAction(Template string, action string) {
+	switch Template {
+	case "rbdDirPath + csiDriverObject":
```

Comment:
huh? how does this ever match?!

I don't think this is ready for review, marking it **draft** for the time being.